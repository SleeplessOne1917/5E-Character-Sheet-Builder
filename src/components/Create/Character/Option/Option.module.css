.chevron {
    height: 3rem;
    width: 3rem;
}

.option-container {
    display: flex;
    justify-content: space-between;
    align-items: center;
    background-color: var(--davy);
    color: var(--heaven);
    padding: 1rem;
    cursor: pointer;
    border-radius: 0.5rem;
}

.option-container:focus-visible {
    border: var(--focus-outline);
    outline: 2px solid var(--davy)
}

.option-title {
    font-family: var(--font-fantasy);
    font-size: 1.5rem;
    display: flex;
    align-items: center;
}

.icon-container {
    height: 4.5rem;
    width: 4.5rem;
    background-color: var(--heaven);
    border-radius: 50%;
    margin-right: 1rem;
    display: flex;
    align-items: center;
    justify-content: center;
}

.icon {
    height: 3.5rem;
    width: 3.5rem;
}

.option-li {
    margin: 1rem 0;
}

.sub-option-list {
    list-style: none;
}

.sub-option-list li {
    margin: 1rem 0;
}

@media screen and (min-width: 768px) {
    .race-title {
        font-size: 2rem;
    }
}