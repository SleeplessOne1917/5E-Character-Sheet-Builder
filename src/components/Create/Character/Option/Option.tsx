'use client';

import {
	ChevronDownIcon,
	ChevronRightIcon,
	ChevronUpIcon
} from '@heroicons/react/24/solid';
import {
	KeyboardEvent,
	KeyboardEventHandler,
	useCallback,
	useState
} from 'react';

import { SrdItem } from '../../../../types/srd';
import classes from './Option.module.css';
import { handleKeyDownEvent } from '../../../../services/handlerService';

type OptionProps = {
	option: SrdItem;
	subOptions?: SrdItem[];
	onChoose: (subOptionIndex?: string) => void;
};

const Option = ({ option, subOptions, onChoose }: OptionProps) => {
	const hasMultipleSubOptions = subOptions && subOptions.length > 1;
	const hasOneSubOption = subOptions && subOptions.length === 1;
	const title = hasOneSubOption ? subOptions[0].name : option.name;
	const [showSubOptions, setShowSubOptions] = useState(false);

	const handleClick = useCallback(() => {
		if (hasMultipleSubOptions) {
			setShowSubOptions(prevState => !prevState);
		} else if (hasOneSubOption) {
			onChoose(subOptions[0].index);
		} else {
			onChoose();
		}
	}, [
		setShowSubOptions,
		hasMultipleSubOptions,
		onChoose,
		hasOneSubOption,
		subOptions
	]);

	const handleKeyDown: KeyboardEventHandler<HTMLDivElement> = useCallback(
		event => {
			handleKeyDownEvent<HTMLDivElement>(event, handleClick);
		},
		[handleClick]
	);

	const handleSubOptionClick = useCallback(
		(subraceIndex: string) => {
			onChoose(subraceIndex);
		},
		[onChoose]
	);

	const handleSubOptionKeyDown = useCallback(
		(event: KeyboardEvent<HTMLDivElement>, subraceIndex: string) => {
			handleKeyDownEvent<HTMLDivElement>(event, () => onChoose(subraceIndex));
		},
		[onChoose]
	);

	return (
		<li className={classes['option-li']} data-testid="option">
			<div>
				<div
					className={classes['option-container']}
					tabIndex={0}
					onClick={handleClick}
					onKeyDown={handleKeyDown}
					aria-label={title}
					role="button"
				>
					<div className={classes['option-title']}>
						<div className={classes['icon-container']}>
							<svg className={classes.icon}>
								<use xlinkHref={`/Icons.svg#${option.index}`} />
							</svg>
						</div>{' '}
						{title}
					</div>
					{hasMultipleSubOptions ? (
						showSubOptions ? (
							<ChevronUpIcon
								className={classes.chevron}
								data-testid="chevron-up"
							/>
						) : (
							<ChevronDownIcon
								className={classes.chevron}
								data-testid="chevron-down"
							/>
						)
					) : (
						<ChevronRightIcon
							className={classes.chevron}
							data-testid="chevron-right"
						/>
					)}
				</div>
				{showSubOptions && (
					<ul className={classes['sub-option-list']}>
						{subOptions?.map(subOption => (
							<li key={subOption.index}>
								<div
									className={classes['option-container']}
									tabIndex={0}
									aria-label={subOption.name}
									onClick={() => handleSubOptionClick(subOption.index)}
									onKeyDown={event =>
										handleSubOptionKeyDown(event, subOption.index)
									}
								>
									<div className={classes['option-title']}>
										{subOption.name}
									</div>
									<ChevronRightIcon className={classes.chevron} />
								</div>
							</li>
						))}
					</ul>
				)}
			</div>
		</li>
	);
};

export default Option;
