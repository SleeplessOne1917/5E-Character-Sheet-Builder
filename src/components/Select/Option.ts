type Option = { value: string | number; label: string };

export default Option;
