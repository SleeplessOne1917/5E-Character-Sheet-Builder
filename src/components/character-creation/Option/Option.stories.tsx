import { ComponentMeta, ComponentStory } from '@storybook/react';
import Option from './Option';

export default {
	title: 'Components/CharacterCreation/Option',
	component: Option,
	argTypes: {
		onChoose: { type: 'function' }
	},
	args: {
		iconId: 'dwarf'
	},
	decorators: [story => <ul style={{ listStyle: 'none' }}>{story()}</ul>]
} as ComponentMeta<typeof Option>;

const Template: ComponentStory<typeof Option> = args => <Option {...args} />;

export const OnlyOption = Template.bind({});
OnlyOption.args = {
	option: {
		index: 'dwarf',
		name: 'Dwarf'
	}
};

export const OptionWithOneSubOption = Template.bind({});
OptionWithOneSubOption.args = {
	option: {
		index: 'dwarf',
		name: 'Dwarf'
	},
	subOptions: [
		{
			index: 'hill-dwarf',
			name: 'Hill Dwarf'
		}
	]
};

export const OptionWithMultipleSubOptions = Template.bind({});
OptionWithMultipleSubOptions.args = {
	option: {
		index: 'dwarf',
		name: 'Dwarf'
	},
	subOptions: [
		{
			index: 'hill-dwarf',
			name: 'Hill Dwarf'
		},
		{
			index: 'mountain-dwarf',
			name: 'Mountain Dwarf'
		}
	]
};
