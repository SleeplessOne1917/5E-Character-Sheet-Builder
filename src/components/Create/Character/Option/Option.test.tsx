import '@testing-library/jest-dom';

import * as stories from './Option.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/react';
import userEvent from '@testing-library/user-event';

const { OnlyOption, OptionWithMultipleSubOptions, OptionWithOneSubOption } =
	composeStories(stories);

it('renders correctly', () => {
	render(<OnlyOption />);

	expect(screen.getByTestId(/option/i)).toMatchSnapshot();
});

describe('has correct chevron', () => {
	it('when only race', () => {
		render(<OnlyOption />);

		expect(screen.getByTestId(/chevron-right/i)).toBeInTheDocument();
		expect(screen.queryByTestId(/chevron-up/i)).not.toBeInTheDocument();
		expect(screen.queryByTestId(/chevron-down/i)).not.toBeInTheDocument();
	});

	it('when only one subrace', () => {
		render(<OptionWithOneSubOption />);

		expect(screen.getByTestId(/chevron-right/i)).toBeInTheDocument();
		expect(screen.queryByTestId(/chevron-up/i)).not.toBeInTheDocument();
		expect(screen.queryByTestId(/chevron-down/i)).not.toBeInTheDocument();
	});

	it('when multiple subraces collapsed', () => {
		render(<OptionWithMultipleSubOptions />);

		expect(screen.queryByTestId(/chevron-right/i)).not.toBeInTheDocument();
		expect(screen.queryByTestId(/chevron-up/i)).not.toBeInTheDocument();
		expect(screen.getByTestId(/chevron-down/i)).toBeInTheDocument();
	});

	it('when multiple subraces expanded', async () => {
		render(<OptionWithMultipleSubOptions />);

		await userEvent.click(screen.getByText(/Dwarf/i));

		expect(screen.queryByTestId(/chevron-right/i)).not.toBeInTheDocument();
		expect(screen.getByTestId(/chevron-up/i)).toBeInTheDocument();
		expect(screen.queryByTestId(/chevron-down/i)).not.toBeInTheDocument();
	});
});

it('does not call onChoose when race is clicked and there are multiple subraces', async () => {
	const onChooseMock = jest.fn();

	render(<OptionWithMultipleSubOptions onChoose={onChooseMock} />);

	await userEvent.click(screen.getByText(/Dwarf/i));

	expect(onChooseMock).not.toHaveBeenCalled();
});

describe('onChoose', () => {
	describe('click', () => {
		it('is called with no arguments when no subraces', async () => {
			const onChooseMock = jest.fn();

			render(<OnlyOption onChoose={onChooseMock} />);
			await userEvent.click(screen.getByText(/Dwarf/i));

			expect(onChooseMock).toHaveBeenCalledWith();
		});

		it('is called with subrace when only 1 subrace', async () => {
			const onChooseMock = jest.fn();

			render(<OptionWithOneSubOption onChoose={onChooseMock} />);
			await userEvent.click(screen.getByText(/Dwarf/i));

			expect(onChooseMock).toHaveBeenCalledWith('hill-dwarf');
		});

		it('is called with subrace when multiple subraces', async () => {
			const onChooseMock = jest.fn();

			render(<OptionWithMultipleSubOptions onChoose={onChooseMock} />);
			await userEvent.click(screen.getByText(/Dwarf/i));
			await userEvent.click(screen.getByText(/Mountain Dwarf/i));

			expect(onChooseMock).toHaveBeenCalledWith('mountain-dwarf');
		});
	});

	describe('on enter press', () => {
		it('is called with no arguments when no subraces', async () => {
			const onChooseMock = jest.fn();

			render(<OnlyOption onChoose={onChooseMock} />);
			await userEvent.tab();
			await userEvent.keyboard('{Enter}');

			expect(onChooseMock).toHaveBeenCalledWith();
		});

		it('is called with subrace when only 1 subrace', async () => {
			const onChooseMock = jest.fn();

			render(<OptionWithOneSubOption onChoose={onChooseMock} />);
			await userEvent.tab();
			await userEvent.keyboard('{Enter}');

			expect(onChooseMock).toHaveBeenCalledWith('hill-dwarf');
		});

		it('is called with subrace when multiple subraces', async () => {
			const onChooseMock = jest.fn();

			render(<OptionWithMultipleSubOptions onChoose={onChooseMock} />);
			await userEvent.tab();
			await userEvent.keyboard('{Enter}');
			await userEvent.tab();
			await userEvent.tab();
			await userEvent.keyboard('{Enter}');

			expect(onChooseMock).toHaveBeenCalledWith('mountain-dwarf');
		});
	});

	describe('on space press', () => {
		it('is called with no arguments when no subraces', async () => {
			const onChooseMock = jest.fn();

			render(<OnlyOption onChoose={onChooseMock} />);
			await userEvent.tab();
			await userEvent.keyboard(' ');

			expect(onChooseMock).toHaveBeenCalledWith();
		});

		it('is called with subrace when only 1 subrace', async () => {
			const onChooseMock = jest.fn();

			render(<OptionWithOneSubOption onChoose={onChooseMock} />);
			await userEvent.tab();
			await userEvent.keyboard(' ');

			expect(onChooseMock).toHaveBeenCalledWith('hill-dwarf');
		});

		it('is called with subrace when multiple subraces', async () => {
			const onChooseMock = jest.fn();

			render(<OptionWithMultipleSubOptions onChoose={onChooseMock} />);
			await userEvent.tab();
			await userEvent.keyboard(' ');
			await userEvent.tab();
			await userEvent.tab();
			await userEvent.keyboard(' ');

			expect(onChooseMock).toHaveBeenCalledWith('mountain-dwarf');
		});
	});
});
