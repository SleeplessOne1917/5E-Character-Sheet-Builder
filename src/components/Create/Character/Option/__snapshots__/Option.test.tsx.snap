// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<li
  class="option-li"
  data-testid="option"
>
  <div>
    <div
      aria-label="Dwarf"
      class="option-container"
      role="button"
      tabindex="0"
    >
      <div
        class="option-title"
      >
        <div
          class="icon-container"
        >
          <svg
            class="icon"
          >
            <use
              xlink:href="/Icons.svg#dwarf"
            />
          </svg>
        </div>
         
        Dwarf
      </div>
      <svg
        aria-hidden="true"
        class="chevron"
        data-testid="chevron-right"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M16.28 11.47a.75.75 0 010 1.06l-7.5 7.5a.75.75 0 01-1.06-1.06L14.69 12 7.72 5.03a.75.75 0 011.06-1.06l7.5 7.5z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</li>
`;
